src/statistics/Statistics.ts

```diff
@@ -167,31 +190,85 @@ export class Statistics {
 
         stats.push(name);
         stats.push(Statistics.KEY_VALUE_SEPARATOR);
-        stats.push(value);
+        if (value === null) {
+            stats.push(Statistics.EMPTY_STAT_VALUE);
+        } else {
+            stats.push('' + value);
+        }
     }
 
-    private addEmptyStat(stats: string[], name: string, keyPrefix: string): void {
-        this.addStat(stats, name, Statistics.EMPTY_STAT_VALUE, keyPrefix);
+    private addEmptyAttribute(stats: string[],
+                              name: string,
+                              keyPrefix: string): void {
+        this.addAttribute(stats, name, Statistics.EMPTY_STAT_VALUE, keyPrefix);
     }
 
-    private fillMetrics(stats: string[], connection: ClientConnection): void {
-        this.addStat(stats, 'lastStatisticsCollectionTime', Date.now());
-        this.addStat(stats, 'enterprise', 'false');
-        this.addStat(stats, 'clientType', this.client.getClusterService().getLocalClient().type);
-        this.addStat(stats, 'clientVersion', BuildInfo.getClientVersion());
-        this.addStat(stats, 'clusterConnectionTimestamp', connection.getStartTime());
-        this.addStat(stats, 'clientAddress', connection.getLocalAddress().toString());
-        this.addStat(stats, 'clientName', this.client.getName());
+    private addMetric(compressor: MetricsCompressor,
+                      descriptor: MetricDescriptor,
+                      value: number,
+                      type: ValueType): void {
+        try {
+            switch (type) {
+                case ValueType.LONG:
+                    compressor.addLong(descriptor, value);
+                    break;
+                case ValueType.DOUBLE:
+                    compressor.addDouble(descriptor, value);
+                    break;
+                default:
+                    throw new Error('Unexpected type: ' + type);
+            }
+        } catch (err) {
+            this.logCompressorError(err);
+        }
+    }
 
-        for (const gaugeName in this.allGauges) {
-            const gaugeValueFunc = this.allGauges[gaugeName];
+    private addSimpleMetric(compressor: MetricsCompressor,
+                            metric: string,
+                            value: number,
+                            type: ValueType,
+                            unit?: ProbeUnit): void {
+        const dotIdx = metric.lastIndexOf('.');
```

Comment:
Maybe not a big deal since the gauge names are pretty short, but maybe we can separate them in the registration part and get rid of this index checks are substring operations. `this.registerGauge('runtime.maxMemory', ...` -> `this.registerGauge('runtime', 'maxMemory', ...`

Reply:
I'd prefer to keep them as is as current approach makes registration and gauge handling more readable. These `String.lastIndexOf()` should be pretty cheap especially when compared with the subsequent zlib compression.

Reply:
Ok, sounds good to me